src/snowflake/connector/s3_storage_client.py

```diff
@@ -87,84 +92,145 @@ def __init__(
             )
         else:
             if self.use_s3_regional_url:
-                self.endpoint = f"https://{self.s3location.bucket_name}.s3.{self.region_name}.amazonaws.com"
+                self.endpoint = (
+                    f"https://{self.s3location.bucket_name}."
+                    f"s3.{self.region_name}.amazonaws.com"
+                )
             else:
                 self.endpoint = (
                     f"https://{self.s3location.bucket_name}.s3.amazonaws.com"
                 )
 
     @staticmethod
-    def sign(secret_key: bytes, msg: bytes) -> bytes:
-        h = hmac.HMAC(secret_key, hashes.SHA1())
-        h.update(msg)
-        return base64.encodebytes(h.finalize()).strip()
+    def _sign_bytes(secret_key: bytes, _input: str) -> bytes:
+        """Applies HMAC-SHA-256 to given string with secret_key."""
+        h = hmac.HMAC(secret_key, hashes.SHA256())
+        h.update(_input.encode("utf-8"))
+        return h.finalize()
 
     @staticmethod
-    def _construct_canonicalized_element(
-        bucket_name: Optional[str] = None,
-        request_uri: str = "",
-        subresource: Optional[Dict[str, Union[str, int, None]]] = None,
+    def _sign_bytes_hex(secret_key: bytes, _input: str) -> bytes:
+        """Convenience function, same as _sign_bytes, but returns result in hex form."""
+        return binascii.hexlify(SnowflakeS3RestClient._sign_bytes(secret_key, _input))
+
+    @staticmethod
+    def _hash_bytes(_input: bytes) -> bytes:
+        """Applies SHA-256 hash to given bytes."""
+        digest = hashes.Hash(hashes.SHA256())
+        digest.update(_input)
+        return digest.finalize()
+
+    @staticmethod
+    def _hash_bytes_hex(_input: bytes) -> bytes:
+        """Convenience function, same as _hash_bytes, but returns result in hex form."""
+        return binascii.hexlify(SnowflakeS3RestClient._hash_bytes(_input))
+
+    @staticmethod
+    def _construct_query_string(
+        query_parts: Tuple[Tuple[str, str], ...],
     ) -> str:
```

Comment:
This function simply turns an input of the form `(("key1", "valu1"), ("key2", "value2"),)` to the string `key1=value1&key2=value2`.
It also works with simple end-points like accelerate where `(("accelerate", ""),)` turns into `accelerate`